File: bindings/python/src/kortecx/tools.py
"""Local function tools — the ``@kx.tool`` decorator (Batch V2b).

```python
import kortecx as kx

@kx.tool
def add(a: int, b: int) -> int:
    "Add two integers."
    return a + b

# fire it deterministically (works today):
print(kx.flow().tool(add, a=2, b=2).run().text)

# or let a model decide (the steered react-auto lane; needs KX_SERVE_AUTOGRANT=1):
print(kx.Agent("Do the math.", tools=[add], dynamic=True).run("what is 2+2?").text)
```

Decorating a Python function turns it into a real, governed, fireable tool with
**zero new runtime substrate**: the SDK exposes the decorated functions as a local
**stdio MCP server** (:mod:`kortecx._toolserver`) and the runtime DIALS it through
the existing PR-6b MCP gateway (``RegisterMcpServer`` / ``DiscoverServerTools``).
A local tool is just another external MCP tool the runtime fires under a
server-built warrant (SN-8 — the client never supplies a warrant).

**Dev-scoped by design.** The runtime spawns the stdio server subprocess, so the
runtime, this interpreter, and the tool module must be co-located (the SDK user is
the operator on their own machine). Registering a stdio MCP server is the same
host-trusted operation as ``kx connections add --command`` (PR-6b/D81/GR19); V2b
adds no new attack surface. Cloud governs the bridge.

**Three firing lanes (D161, GR15-honest):**

- **deterministic** — ``flow().tool(fn, **args)`` fires ONE tool as a standalone
  node (works today, even model-free).
- **steered / dynamic** — ``Agent(tools=[fn], dynamic=True)`` runs the
  ``kx/recipes/react-auto`` loop where the model picks tools (needs a served model
  + ``KX_SERVE_AUTOGRANT=1``).
- **frozen / deterministic-agentic** — ``Agent(tools=[fn])`` (the default lane) is a
  MODEL step with a fixed tool set; its bounded loop **lands in PR-9b-2** and is
  refused at submit today, so the Agent raises a clear pre-flight hint until then.

**Re-import contract.** The runtime spawns ``python -m kortecx._toolserver`` which
re-loads your module to recover the functions. Decorate at MODULE level and guard
any ``.run()`` calls under ``if __name__ == "__main__":`` (the toolserver loads the
module under a non-``__main__`` name, so a guarded main block never re-runs).
"""

from __future__ import annotations

import enum
import inspect
import os
import sys
import typing
import warnings
from dataclasses import dataclass
from typing import (
    TYPE_CHECKING,
    Any,
    Callable,
    Dict,
    List,
    Mapping,
    Optional,
    Sequence,
    Tuple,
    Union,
)

from .blueprints import TOOL_ARGS_KEY, StepInput

if TYPE_CHECKING:
    from .chains import Chain, Task

#: Env sentinel set by :mod:`kortecx._toolserver` while it re-imports the user's
#: module. The lazy default client refuses to run a flow/agent in this context
#: (a guard against an un-guarded top-level ``.run()`` re-executing + recursing).
TOOLSERVE_ENV = "KX_TOOLSERVE"

#: The stable server-name prefix for an SDK-registered local stdio MCP server.
_LOCAL_SERVER_PREFIX = "kxlocal-"


@dataclass(eq=False)
class LocalToolDef:
    """A function exposed as a local MCP tool (identity = the object). ``fn`` runs
    IN the toolserver subprocess; ``schema`` is the derived MCP ``inputSchema``."""

    name: str
    version: str
    description: str
    schema: Dict[str, Any]
    fn: Callable[..., Any]
    module: str
    qualname: str
    script_path: str


#: Process-global registry the toolserver subprocess reads after re-importing a
#: module (name → def). In the authoring process it simply accumulates decorations.
_LOCAL_TOOLS: Dict[str, LocalToolDef] = {}


# --- type-hint → MCP inputSchema (the gateway's mapper gates str/int/bool/enum) ---


def _type_to_jsonschema(name: str, ann: Any) -> Dict[str, Any]:
    """Map ONE parameter annotation to a JSON-Schema property. Mirrors what the
    gateway's ``json_schema_to_input_schema`` actually type-gates: ``str`` /
    ``int`` / ``bool`` / string-enum. ``float`` maps to ``number`` (the runtime
    does NOT gate floats — a warn mirrors the ``tool()`` no-floats rule); any
    other / missing annotation falls back to ``string`` (the arg passes verbatim)."""
    origin = typing.get_origin(ann)
    if origin is typing.Literal:
        vals = list(typing.get_args(ann))
        if vals and all(isinstance(v, str) for v in vals):
            return {"enum": vals}
        return {"type": "string"}
    if isinstance(ann, type) and issubclass(ann, enum.Enum):
        members = [m.value for m in ann]
        if members and all(isinstance(v, str) for v in members):
            return {"enum": members}
        return {"type": "string"}
    # NOTE: bool is a subclass of int — check it first.
    if ann is bool:
        return {"type": "boolean"}
    if ann is int:
        return {"type": "integer"}
    if ann is str:
        return {"type": "string"}
    if ann is float:
        warnings.warn(
            f"@kx.tool: parameter {name!r} is typed float; the runtime does not "
            "type-gate numbers — it is passed verbatim. Prefer int where possible.",
            stacklevel=4,
        )
        return {"type": "number"}
    # Unknown/complex (list/dict/typed object) — no client gate; document that the
    # function receives the decoded JSON value verbatim.
    return {"type": "string"}


def _schema_from_signature(fn: Callable[..., Any]) -> Dict[str, Any]:
    """Derive an MCP ``inputSchema`` (``{type:object, properties, required}``) from
    ``fn``'s signature + type hints. Params without a default are ``required``;
    ``*args`` / ``**kwargs`` are skipped (un-mappable — documented)."""
    sig = inspect.signature(fn)
    try:
        hints = typing.get_type_hints(fn)
    except Exception:  # pragma: no cover - exotic annotations
        hints = {}
    props: Dict[str, Any] = {}
    required: List[str] = []
    for pname, param in sig.parameters.items():
        if param.kind in (param.VAR_POSITIONAL, param.VAR_KEYWORD):
            continue
        ann = hints.get(pname, param.annotation)
        props[pname] = _type_to_jsonschema(pname, ann)
        if param.default is inspect.Parameter.empty:
            required.append(pname)
    schema: Dict[str, Any] = {"type": "object", "properties": props}
    if required:
        schema["required"] = required
    return schema


def _script_path_of(fn: Callable[..., Any]) -> str:
    """The abs path of the file defining ``fn`` (the toolserver re-loads it). Raise
    a clear error for a function with no source file (a REPL/notebook closure)."""
    src = inspect.getsourcefile(fn) or inspect.getfile(fn)
    if not src:
        raise ToolError(
            f"@kx.tool: cannot locate the source file for {getattr(fn, '__name__', fn)!r}; "
            "a local tool must be a function defined in a .py file (not a REPL closure)."
        )
    return os.path.abspath(src)


class ToolError(ValueError):
    """A local-tool authoring/registration error."""


def _make_local_tool(
    fn: Callable[..., Any],
    *,
    name: Optional[str] = None,
    version: str = "1",
    description: Optional[str] = None,
) -> Callable[..., Any]:
    """Tag ``fn`` with a :class:`LocalToolDef` and register it. Returns ``fn``
    unchanged (still directly callable)."""
    if not callable(fn):
        raise ToolError("@kx.tool must decorate a callable")
    tool_name = name or fn.__name__.replace("_", "-")
    desc = description if description is not None else (inspect.getdoc(fn) or "")
    tdef = LocalToolDef(
        name=tool_name,
        version=str(version),
        description=desc.strip(),
        schema=_schema_from_signature(fn),
        fn=fn,
        module=getattr(fn, "__module__", ""),
        qualname=getattr(fn, "__qualname__", tool_name),
        script_path=_script_path_of(fn),
    )
    _LOCAL_TOOLS[tool_name] = tdef
    fn.__kx_tool__ = tdef  # type: ignore[attr-defined]
    return fn


def local_tool_def(fn: Any) -> Optional[LocalToolDef]:
    """Return the :class:`LocalToolDef` a ``@kx.tool``-decorated ``fn`` carries, or
    ``None`` (not a local tool)."""
    return getattr(fn, "__kx_tool__", None)


# --- the public `tool` symbol: a decorator AND the back-compat node factory -------


@typing.overload
def tool(fn: Callable[..., Any]) -> Callable[..., Any]: ...


@typing.overload
def tool(
    *, name: Optional[str] = ..., version: str = ..., description: Optional[str] = ...
) -> Callable[[Callable[..., Any]], Callable[..., Any]]: ...


@typing.overload
def tool(tool_id: str, tool_version: str, **args: object) -> "Task": ...


def tool(*args: Any, **kwargs: Any) -> Any:
    """``@kx.tool`` — expose a local function as a governed tool (V2b); OR
    ``tool("id", "version", **args)`` — a standalone TOOL node firing a REGISTERED
    tool (PR-6b-2, the back-compat factory).

    Decorator forms::

        @kx.tool
        def add(a: int, b: int) -> int: ...

        @kx.tool(name="adder", version="2")
        def add(a: int, b: int) -> int: ...

    The function's type hints become the tool's MCP ``inputSchema`` (str / int /
    bool / string-enum are type-gated by the runtime; floats pass verbatim).
    """
    # Bare decorator: @kx.tool
    if len(args) == 1 and callable(args[0]) and not kwargs:
        return _make_local_tool(args[0])
    # Node factory (back-compat): tool("id", "version", **args)
    if args and isinstance(args[0], str):
        from .chains import tool as _node_tool

        return _node_tool(*args, **kwargs)
    # Parametrized decorator: @kx.tool(name=..., version=..., description=...)
    if not args:
        name = kwargs.pop("name", None)
        version = kwargs.pop("version", "1")
        description = kwargs.pop("description", None)
        if kwargs:
            raise ToolError(f"@kx.tool got unexpected keyword(s): {sorted(kwargs)}")

        def _decorate(fn: Callable[..., Any]) -> Callable[..., Any]:
            return _make_local_tool(fn, name=name, version=version, description=description)

        return _decorate
    raise ToolError(
        "tool(...) takes either @kx.tool on a function, @kx.tool(name=..., version=...), "
        'or the node factory tool("id", "version", **args)'
    )


# --- run-terminal resolution: register the stdio server(s) + name the tools -------

ToolsArg = Optional[Union[Sequence[Any], Mapping[str, str]]]


def split_tools(tools: ToolsArg) -> Tuple[ToolsArg, Tuple[LocalToolDef, ...]]:
    """Split a ``tools=`` value into (string/mapping grants, local-tool defs). A bare
    function must be ``@kx.tool``-decorated; a plain lambda is a fail-closed error."""
    if tools is None or isinstance(tools, Mapping):
        return tools, ()
    names: List[Any] = []
    locals_: List[LocalToolDef] = []
    for t in tools:
        tdef = local_tool_def(t)
        if tdef is not None:
            locals_.append(tdef)
        elif callable(t):
            raise ToolError(
                "a plain function in tools=[...] must be decorated with @kx.tool "
                f"(got {getattr(t, '__name__', t)!r})"
            )
        else:
            names.append(t)
    return (names if names else None), tuple(locals_)


def local_tool_node(tdef: LocalToolDef, args: Mapping[str, object]) -> "Task":
    """A standalone TOOL node for a local function — the tool_contract is filled at
    resolution (the namespaced ``<server>/<name>`` is server-derived)."""
    from .chains import Task, _canonical_args_json

    return Task(
        StepInput(
            kind="tool",
            tool_contract={},
            params={TOOL_ARGS_KEY: _canonical_args_json(dict(args))},
            local_tools=(tdef,),
        )
    )


def _server_name_for(script_path: str) -> str:
    """A stable, deterministic server name per defining script (so re-runs upsert
    the same connection — ``connection_id_of(name)`` is deterministic, SN-8)."""
    import hashlib

    digest = hashlib.blake2b(script_path.encode("utf-8"), digest_size=8).hexdigest()
    return f"{_LOCAL_SERVER_PREFIX}{digest}"


def _server_args(script_path: str, names: Sequence[str]) -> List[str]:
    return [
        "-m",
        "kortecx._toolserver",
        "--script",
        script_path,
        "--tools",
        ",".join(sorted(set(names))),
    ]


def _plan_registrations(
    defs: Sequence[LocalToolDef],
) -> List[Tuple[str, str, List[str]]]:
    """Group defs by defining script → one stdio server each. Returns
    ``(server_name, script_path, [tool_names])`` per server, deterministically."""
    by_script: Dict[str, List[LocalToolDef]] = {}
    for d in defs:
        by_script.setdefault(d.script_path, []).append(d)
    plan: List[Tuple[str, str, List[str]]] = []
    for script in sorted(by_script):
        names = [d.name for d in by_script[script]]
        plan.append((_server_name_for(script), script, names))
    return plan


def _resolved_name(server_name: str, tool_name: str, registered: Sequence[Any]) -> str:
    """Find the namespaced ``<server>/<name>`` for ``tool_name`` among a server's
    discovered tools (read back — never guess the namespacing rule)."""
    want = f"{server_name}/{tool_name}"
    for rt in registered:
        if getattr(rt, "tool_name", None) == want:
            return rt.tool_name
    # Fall back to a remote-segment match (defensive against any normalization).
    for rt in registered:
        full = getattr(rt, "tool_name", "")
        if "/" in full and full.split("/", 1)[1] == tool_name:
            return full
    raise ToolError(
        f"local tool {tool_name!r} was not discovered on server {server_name!r} "
        "(registration/discovery mismatch)"
    )


def _apply_contract(chain: "Chain", name_map: Dict[int, str]) -> None:
    """Augment each step's ``tool_contract`` with its resolved local-tool names
    (idempotent ``setdefault`` — safe to re-run)."""
    for step in chain._iter_steps():
        for tdef in step.local_tools:
            resolved = name_map.get(id(tdef))
            if resolved is not None:
                step.tool_contract.setdefault(resolved, tdef.version)


def resolve_local_tools(client: Any, chain: "Chain") -> None:
    """Register every local tool referenced by ``chain`` (sync) + rewrite each
    step's ``tool_contract`` to the server-derived names. No-op when there are none."""
    defs = _collect_defs(chain)
    if not defs:
        return
    by_def: Dict[str, List[LocalToolDef]] = {}
    for d in defs:
        by_def.setdefault(d.script_path, []).append(d)
    name_map: Dict[int, str] = {}
    for server_name, script, names in _plan_registrations(defs):
        client.register_mcp_server(
            name=server_name,
            transport="stdio",
            endpoint=sys.executable,
            args=_server_args(script, names),
        )
        page = client.discover_server_tools(name=server_name)
        for tdef in by_def[script]:
            name_map[id(tdef)] = _resolved_name(server_name, tdef.name, page.tools)
    _apply_contract(chain, name_map)


async def aresolve_local_tools(client: Any, chain: "Chain") -> None:
    """As :func:`resolve_local_tools`, over an async client."""
    defs = _collect_defs(chain)
    if not defs:
        return
    by_def: Dict[str, List[LocalToolDef]] = {}
    for d in defs:
        by_def.setdefault(d.script_path, []).append(d)
    name_map: Dict[int, str] = {}
    for server_name, script, names in _plan_registrations(defs):
        await client.register_mcp_server(
            name=server_name,
            transport="stdio",
            endpoint=sys.executable,
            args=_server_args(script, names),
        )
        page = await client.discover_server_tools(name=server_name)
        for tdef in by_def[script]:
            name_map[id(tdef)] = _resolved_name(server_name, tdef.name, page.tools)
    _apply_contract(chain, name_map)


def _collect_defs(chain: "Chain") -> List[LocalToolDef]:
    seen: Dict[int, LocalToolDef] = {}
    for step in chain._iter_steps():
        for tdef in step.local_tools:
            seen.setdefault(id(tdef), tdef)
    return list(seen.values())


def register_tools(client: Any, tools: ToolsArg) -> None:
    """Register the local tools in a ``tools=`` set (sync) WITHOUT rewriting a
    contract — used by the dynamic react-auto lane, which auto-grants the live
    registry. No-op when there are no local tools."""
    _, defs = split_tools(tools)
    if not defs:
        return
    for server_name, script, names in _plan_registrations(defs):
        client.register_mcp_server(
            name=server_name,
            transport="stdio",
            endpoint=sys.executable,
            args=_server_args(script, names),
        )


async def aregister_tools(client: Any, tools: ToolsArg) -> None:
    """As :func:`register_tools`, over an async client."""
    _, defs = split_tools(tools)
    if not defs:
        return
    for server_name, script, names in _plan_registrations(defs):
        await client.register_mcp_server(
            name=server_name,
            transport="stdio",
            endpoint=sys.executable,
            args=_server_args(script, names),
        )

File: bindings/typescript/src/tools.ts
/**
 * Local function tools — `localTool(...)` (Batch V2b), the TS mirror of Python's
 * `@kx.tool`.
 *
 * ```ts
 * import { localTool, flow } from "@kortecx/sdk";
 *
 * const add = localTool({
 *   name: "add",
 *   params: { a: "integer", b: "integer" }, // explicit — TS types are erased at runtime
 *   run: ({ a, b }) => a + b,
 * });
 *
 * await flow().tool(add, { a: 2, b: 2 }).run({ client: kx }); // deterministic, today
 * ```
 *
 * `localTool` exposes a JS function as a real, governed tool: the SDK runs it as a
 * local **stdio MCP server** (`_toolserver`) the runtime DIALS through the existing
 * PR-6b MCP gateway — zero new runtime substrate; the runtime fires it under a
 * server-built warrant (SN-8).
 *
 * **Node + dev-scoped.** The runtime spawns the stdio server subprocess, so this is
 * the Node SDK only (a browser cannot spawn a subprocess) and the tool MODULE must
 * be Node-importable (compiled JS / `.mjs`) and co-located with the serve. Unlike
 * Python (type hints → schema), TS types are erased, so the param schema is explicit.
 *
 * **Firing lanes (GR15-honest):** deterministic `flow().tool(fn, args)` (today) ·
 * steered `new Agent({ tools:[fn], dynamic:true })` → `kx/recipes/react-auto` (today,
 * needs `KX_SERVE_AUTOGRANT=1`) · frozen agentic loop → PR-9b-2 (a clear pre-flight hint).
 */

import { type Chain, type Task, task } from "./chains.js";

/** A local-tool authoring/registration error. */
export class KxToolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "KxToolError";
  }
}

/** A param's declared type — `"string"` / `"integer"` / `"boolean"`, or a string
 *  array (an exact-match enum). Floats are intentionally absent (the runtime does
 *  not type-gate numbers; pass an int). */
export type LocalParamType = "string" | "integer" | "boolean" | readonly string[];

/** A param spec: a bare type (required) or `{ type, required? }`. */
export type LocalParamSpec = LocalParamType | { type: LocalParamType; required?: boolean };

/** The spec passed to {@link localTool}. */
export interface LocalToolSpec {
  /** The tool name (the model + `flow().tool` reference it). */
  name: string;
  /** The argument schema (explicit — TS types are erased at runtime). */
  params?: Readonly<Record<string, LocalParamSpec>>;
  /** The implementation — receives the decoded args, returns a JSON-able value. */
  run: (args: Record<string, unknown>) => unknown;
  version?: string;
  description?: string;
  /** The defining module's `import.meta.url` (auto-captured from the call site if
   *  omitted; pass it explicitly if capture fails, e.g. under a bundler). */
  module?: string;
}

/** A function exposed as a local MCP tool (the value put in `tools: [...]`). */
export interface LocalToolDef {
  readonly __kxLocalTool: true;
  readonly name: string;
  readonly version: string;
  readonly description: string;
  /** The derived MCP `inputSchema`. */
  readonly schema: Record<string, unknown>;
  /** The implementation (run IN the toolserver subprocess). */
  readonly run: (args: Record<string, unknown>) => unknown;
  /** The defining module (file URL / path), used to re-import in the toolserver. */
  readonly module: string;
}

/** Process-global registry the toolserver subprocess reads after re-importing a
 *  module (name → def). Backed by `globalThis` so it is ONE Map even when `tools.ts`
 *  is inlined into multiple bundles (tsup `splitting: false`) — the user registers via
 *  the SDK bundle, the toolserver reads via its own; both must see the same registry. */
export const LOCAL_TOOLS: Map<string, LocalToolDef> = (() => {
  const key = "__kx_local_tools__";
  const g = globalThis as Record<string, unknown>;
  const existing = g[key];
  if (existing instanceof Map) {
    return existing as Map<string, LocalToolDef>;
  }
  const m = new Map<string, LocalToolDef>();
  g[key] = m;
  return m;
})();

function jsonProp(t: LocalParamType): Record<string, unknown> {
  if (Array.isArray(t)) {
    return { enum: [...t] };
  }
  if (t === "integer") {
    return { type: "integer" };
  }
  if (t === "boolean") {
    return { type: "boolean" };
  }
  return { type: "string" };
}

function buildSchema(params?: Readonly<Record<string, LocalParamSpec>>): Record<string, unknown> {
  const properties: Record<string, unknown> = {};
  const required: string[] = [];
  for (const [name, spec] of Object.entries(params ?? {})) {
    if (typeof spec === "string" || Array.isArray(spec)) {
      properties[name] = jsonProp(spec); // a bare type or a string-enum array
      required.push(name);
    } else if (typeof spec === "object" && "type" in spec) {
      properties[name] = jsonProp(spec.type);
      if (spec.required !== false) {
        required.push(name);
      }
    }
  }
  const schema: Record<string, unknown> = { type: "object", properties };
  if (required.length > 0) {
    schema.required = required;
  }
  return schema;
}

/** Best-effort caller-module capture from the stack (no Node import — browser-safe).
 *  The first stack frame is THIS file (the SDK bundle — `dist/node.js` etc. once
 *  bundled, not literally `tools.ts`), so we skip every frame sharing that file and
 *  return the first user frame. Robust to bundling (tsup inlines `tools.ts` into each
 *  entry, so a filename-based skip would miss it). */
function callerModule(): string {
  const stack = new Error().stack ?? "";
  let selfFile: string | undefined;
  for (const line of stack.split("\n").slice(1)) {
    const m = line.match(/(file:\/\/\/\S+?|[/\\]\S+?):\d+:\d+/);
    if (m === null || m[1] === undefined) {
      continue;
    }
    const p: string = m[1];
    if (selfFile === undefined) {
      // The frame of `callerModule` itself — the SDK bundle to skip past.
      selfFile = p;
      continue;
    }
    if (p === selfFile || p.includes("/node_modules/")) {
      continue;
    }
    return p;
  }
  throw new KxToolError(
    "could not determine the calling module for localTool(); pass { module: import.meta.url }",
  );
}

/** Expose a JS function as a governed local tool (V2b). Returns a {@link LocalToolDef}
 *  to put in `tools: [...]` or `flow().tool(def, args)`; also registers it (so the
 *  toolserver subprocess recovers it after re-importing the module). */
export function localTool(spec: LocalToolSpec): LocalToolDef {
  if (typeof spec.run !== "function") {
    throw new KxToolError("localTool({ run }) must be a function");
  }
  const def: LocalToolDef = {
    __kxLocalTool: true,
    name: spec.name,
    version: spec.version ?? "1",
    description: spec.description ?? "",
    schema: buildSchema(spec.params),
    run: spec.run,
    module: spec.module ?? callerModule(),
  };
  LOCAL_TOOLS.set(def.name, def);
  return def;
}

/** Is `t` a {@link LocalToolDef}? */
export function isLocalTool(t: unknown): t is LocalToolDef {
  return (
    typeof t === "object" && t !== null && (t as { __kxLocalTool?: boolean }).__kxLocalTool === true
  );
}

/** The local-tool defs in a `tools` value (for the dynamic react-auto lane). */
export function localToolsOf(
  tools?: readonly (string | LocalToolDef)[] | Readonly<Record<string, string>>,
): LocalToolDef[] {
  if (tools === undefined || !Array.isArray(tools)) {
    return [];
  }
  return tools.filter(isLocalTool);
}

/** A deterministic dependency-free server name per defining module (so re-runs
 *  upsert the same connection — `connection_id_of(name)` is deterministic, SN-8). */
export function serverNameFor(module: string): string {
  // FNV-1a 32-bit — stable + dependency-free (no cryptographic strength needed).
  let h = 0x811c9dc5;
  for (let i = 0; i < module.length; i++) {
    h ^= module.charCodeAt(i);
    h = Math.imul(h, 0x01000193);
  }
  return `kxlocal-${(h >>> 0).toString(16).padStart(8, "0")}`;
}

/** A minimal client surface the resolver needs. */
export interface ToolGatewayClient {
  registerMcpServer(input: {
    name: string;
    transport: string;
    endpoint: string;
    args: string[];
  }): Promise<unknown>;
  discoverServerTools(name: string): Promise<{ tools: ReadonlyArray<{ toolName: string }> }>;
}

/** Resolve a local tool's namespaced `<server>/<name>` from a server's discovered set. */
function resolvedName(
  serverName: string,
  toolName: string,
  registered: ReadonlyArray<{ toolName: string }>,
): string {
  const want = `${serverName}/${toolName}`;
  for (const rt of registered) {
    if (rt.toolName === want) {
      return rt.toolName;
    }
  }
  for (const rt of registered) {
    const full = rt.toolName;
    if (full.includes("/") && full.slice(full.indexOf("/") + 1) === toolName) {
      return full;
    }
  }
  throw new KxToolError(
    `local tool '${toolName}' was not discovered on server '${serverName}' (registration mismatch)`,
  );
}

/** Guard: refuse to (re)register while the toolserver is re-importing a module — a
 *  stray un-guarded top-level `.run()` would recurse. */
function assertNotReentrant(): void {
  const env = (globalThis as { process?: { env?: Record<string, string | undefined> } }).process
    ?.env;
  if (env?.KX_TOOLSERVE !== undefined) {
    throw new KxToolError(
      "cannot start a run while serving localTool() tools (the toolserver re-import context) — " +
        "guard your .run() calls (e.g. behind a main/entry check) so they do not re-execute.",
    );
  }
}

/** The Node binary + the built `_toolserver.js` entry (Node-only; throws in a browser).
 *  The toolserver sits beside the running bundle in `dist/`; we resolve its directory
 *  from `import.meta.url` (the ESM build) or `__dirname` (the CJS build — esbuild emits
 *  one or the other depending on the consumed format). */
async function spawnTarget(): Promise<{ node: string; entry: string }> {
  assertNotReentrant();
  const proc = (globalThis as { process?: { execPath?: string } }).process;
  if (proc?.execPath === undefined) {
    throw new KxToolError(
      "local tools require the Node SDK (@kortecx/sdk/node) — a browser cannot spawn a tool server",
    );
  }
  const { fileURLToPath } = await import("node:url");
  const { dirname, join } = await import("node:path");
  const metaUrl = (import.meta as { url?: string }).url;
  const baseDir =
    typeof metaUrl === "string" && metaUrl.length > 0
      ? dirname(fileURLToPath(metaUrl))
      : ((globalThis as { __dirname?: string }).__dirname ?? __dirname);
  return { node: proc.execPath, entry: join(baseDir, "_toolserver.js") };
}

async function toFileUrl(module: string): Promise<string> {
  if (module.startsWith("file:")) {
    return module;
  }
  const { pathToFileURL } = await import("node:url");
  return pathToFileURL(module).href;
}

/** Register every local tool a chain references (Node), returning the def → server-derived
 *  `<server>/<name>` map (or `undefined` when there are none). */
export async function resolveLocalTools(
  client: ToolGatewayClient,
  chain: Chain,
): Promise<ReadonlyMap<LocalToolDef, string> | undefined> {
  assertNotReentrant();
  const defs = chain.collectLocalTools();
  if (defs.length === 0) {
    return undefined;
  }
  const { node, entry } = await spawnTarget();
  const byModule = new Map<string, LocalToolDef[]>();
  for (const d of defs) {
    const list = byModule.get(d.module);
    if (list === undefined) {
      byModule.set(d.module, [d]);
    } else {
      list.push(d);
    }
  }
  const resolved = new Map<LocalToolDef, string>();
  for (const [module, group] of byModule) {
    const serverName = serverNameFor(module);
    const url = await toFileUrl(module);
    const names = [...new Set(group.map((d) => d.name))].sort();
    await client.registerMcpServer({
      name: serverName,
      transport: "stdio",
      endpoint: node,
      args: [entry, "--module", url, "--tools", names.join(",")],
    });
    const page = await client.discoverServerTools(serverName);
    for (const d of group) {
      resolved.set(d, resolvedName(serverName, d.name, page.tools));
    }
  }
  return resolved;
}

/** Register the local tools in a `tools=` set WITHOUT rewriting a contract — used by
 *  the dynamic react-auto lane (which auto-grants the live registry). */
export async function registerLocalTools(
  client: ToolGatewayClient,
  tools?: readonly (string | LocalToolDef)[] | Readonly<Record<string, string>>,
): Promise<void> {
  const defs = localToolsOf(tools);
  if (defs.length === 0) {
    return;
  }
  const { node, entry } = await spawnTarget();
  const byModule = new Map<string, LocalToolDef[]>();
  for (const d of defs) {
    const list = byModule.get(d.module);
    if (list === undefined) {
      byModule.set(d.module, [d]);
    } else {
      list.push(d);
    }
  }
  for (const [module, group] of byModule) {
    const url = await toFileUrl(module);
    const names = [...new Set(group.map((d) => d.name))].sort();
    await client.registerMcpServer({
      name: serverNameFor(module),
      transport: "stdio",
      endpoint: node,
      args: [entry, "--module", url, "--tools", names.join(",")],
    });
  }
}

/** A standalone TOOL node firing a local function deterministically (V2b). */
export function localToolNode(
  def: LocalToolDef,
  args: Readonly<Record<string, string | number | boolean>> = {},
): Task {
  return task.localTool(def, args);
}
